proposals/per-extension-language-preferences.md

```diff
@@ -0,0 +1,234 @@
+
+When the user preferred language is changed to a new different language by `i18n.setCurrentLanguage(code)`, this event is fired.
+Developers use this event to know the language changed, and might do the following:
+
+- Update the content of extension pages in place (without refreshing the page)
+- Reload extension pages:
+  - If the page is stateless, refresh the page directly.
+  - If the page is stateful, save the current state and then refresh the page.
+  - Prompts the user that the language setting has changed, and asks the user if they want to refresh the page immediately.
+- Re-create the extension context menus, since the title of the context menus shoule be updated.
+- Update the action title and badge text if needed.
+
+#### Behavior of `i18n.getMessage()`, manifest and CSS files
+
+When this proposal is implemented, the behavior of these existing functions will change as follows:
+
+- If the preferred language is not set by `i18n.setCurrentLanguage()`, it is consistent with the existing behavior.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, and the extension supports this language, then prioritize using that language.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language, then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.
+
+### Browser built-in language selection menu
+
+Based on the capabilities provided by this proposal, browsers could provide a unified built-in language selection menu for extensions, like Android and iOS.
+Whether to provide the built-in language selection menu and how to implement it is up to the browser to decide. This is just a suggestion.
+
+The built-in language selection menu has the following benefits:
+- Easy for developers: Developers do not need to implement it themselves, they only need to adapt to this proposal.
+- Easy for users: A unified UI makes it easier for users to use, otherwise each extension might be set up differently.
+- It is convenient for developers to develop and test i18n functions by switching the extension language.
+
+### Integrate other i18n implementations (frameworks) with browser.i18n
+
+This proposal can also be integrated with third-party i18n frameworks, as these frameworks typically allow developers to specify the language to be used.
+For example, developers use `i18n.getCurrentLanguage()` and `i18n.setCurrentLanguage()` for the user preferred language, but don't use `i18n.getMessage()`.
+
+```js
+otherI18nFramework.setLanguage(browser.i18n.getCurrentLanguage());
+
+let text = otherI18nFramework.getMessage(key);
+```
+
+### New Permissions
+
+N/A
+
+### Manifest File Changes
+
+If the browser would like to support the built-in language selection menu for extensions, the manifest file should add a new key for developers to opt-in this feature, like the following:
```

Comment:
I think we should chat about this with other browser vendors, and only keep this section if at least one vendor plans to implement it.

Reply:
@oliverdunk If no browser vendor wants to implement something like this, we can offer these APIs in `browser.management` so an extension can create this UI to customise the languages of other extensions.

Reply:
What would a `browser.management` API look like? If it involves rendering new UI, I don't think Chrome would be supportive. I would expect other browsers to feel similarly if they don't plan on implementing a menu.

Reply:
I moved the "Browser built-in language selection menu" to the "Future Work" section in this document. When browser vendors are interested in this, we can discuss it again.

Reply:
@oliverdunk the whole point of adding APIs to `browser.management` would be to prevent the browser having to render UI but let extensions figure this out. imagine something like:
```js
browser.management.setLanguage(extensionId, locale)
```

To let an extension change the language of another extension.

Reply:
Ah, I missed this. I could go either way on if that should be in scope here but that seems reasonable.